src/core/device.c

```diff
@@ -1075,6 +1079,98 @@ bool device_shall_be_bound_by(Unit *device, Unit *u) {
+ * triggered by the kernel and are missed for that reason. To fix that up device_start() is used to
+ * re-trigger the corresponding uevent of the particular device unit.
+ *
+ * The re-trigger is only done in case the device is not following another device, was never in plugged
+ * state before, is in the dead state, but is actually plugged (corresponding sysfs entries are existent).
+ *
+ * Note: This accounts only for device units which are explicitly referenced as dependency by other units
+ *       and are tagged in udev rules (TAG+="systemd").
+ *
+ *       This might only be useful in highly embedded environments, e.g. automotive control units, where
+ *       existent devices and the whole dependency tree are known, the ecosystem is well-defined and startup
+ *       performance matters.
+ *
+ *       For desktop and server environments systemd-udev-trigger.service should still be used.
+ */
+static int device_start(Unit *u) {
+        Device *d = DEVICE(u);
+        _cleanup_(sd_device_unrefp) sd_device *dev = NULL;
+        _cleanup_free_ char *path = NULL;
+        int r;
+
+        assert(d);
+
+        if (device_following(u))
+                return 0;
+
+        if (!(d->state == DEVICE_DEAD))
+                return 0;
+
+        if (d->plugged_once)
+                return 0;
+
+        r = unit_name_to_path(u->id, &path);
+        if (r)
+                return log_unit_warning_errno(u, SYNTHETIC_ERRNO(EBADR),
+                                "Failed to get device path from unit: \"%s\" (%d)", u->id, r);
+
+        if (path_startswith(path, "/sys/")) {
+                r = sd_device_new_from_syspath(&dev, path);
+                if (r < 0)
+                        return log_unit_debug_errno(u, SYNTHETIC_ERRNO(EBADR),
+                                        "Failed to find device in sysfs: \"%s\" (%d)", path, r);
+        } else if (path_startswith(path, "/dev/")) {
+                struct stat st;
+
+                if (stat(path, &st) < 0) {
+                        r = -errno;
+                        return log_unit_debug_errno(u, SYNTHETIC_ERRNO(EBADR),
+                                        "Failed to find device in devtmpfs: \"%s\" (%d)", path, r);
+                }
+
+                r = sd_device_new_from_stat_rdev(&dev, &st);
+                if (r < 0)
+                        return log_unit_warning_errno(u, SYNTHETIC_ERRNO(EBADR),
+                                        "Failed to find corresponding sysfs entry for: \"%s\" (%d)",
+                                        path, r);
+        } else {
+                return log_unit_warning_errno(u, SYNTHETIC_ERRNO(EBADR),
+                                "Device unit name does not evaluate to either a /sys/ or /dev/ path: \
+                                \"%s\" (%d)", path, r);
+        }
+
+        r = sd_device_trigger(dev, SD_DEVICE_ADD);
+        if (r < 0) {
+                bool ignore = IN_SET(r, -ENODEV);
```

Comment:
It is not necessary to use `IN_SET()` macro if you only ignore `-ENODEV`.
But, see udevadm-trigger.c, I think we should also ignore `-ENOENT` here.

Reply:
I think in this case -ENOENT would actually be an error, since we explicitly requested this device by a unit dependency in order to make sure that:
- it existes and enters plugged state soon
- the particular udev rules are processed